File: backend/app/schemas/agent_factory.py
"""Pydantic request/response schemas for the AI Agent Factory API."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

from app.services.agent_factory.spec import AgentSpec


class CompileRequest(BaseModel):
    """POST /factory/compile — one plain-language sentence."""

    one_liner: str = Field(..., min_length=1, description="一句话需求")


class SpecSummary(BaseModel):
    """Row in the spec list."""

    id: str
    name: str
    version: int
    status: str
    one_liner: str


class SpecResponse(BaseModel):
    """A full spec (the five blueprints) plus its DB status/version."""

    id: str
    name: str
    version: int
    status: str
    spec: AgentSpec


class UpdateSpecRequest(BaseModel):
    """PATCH /factory/specs/{id} — human edits to a draft (esp. policies/tools)."""

    spec: AgentSpec


class RunRequest(BaseModel):
    """POST /factory/agents/{id}/run — a task for the assembled child agent."""

    input: str = Field(..., min_length=1, description="给子 Agent 的一句话任务")


class RunStep(BaseModel):
    """One ReAct step in the returned trace."""

    thought: Optional[str] = None
    action: Optional[str] = None
    observation: Optional[str] = None


class RunResponse(BaseModel):
    """Result of running an assembled child agent."""

    spec_id: str
    final: str
    trace: list[RunStep] = Field(default_factory=list)

File: frontend/src/types/agentFactory.ts
/**
 * AI Agent Factory (孵化器) types — mirror of backend AgentSpec + factory schemas.
 * Backend: app/services/agent_factory/spec.py + app/schemas/agent_factory.py
 */

export type SpecStatus = 'draft' | 'approved' | 'published'
export type ToolRisk = 'low' | 'high'

export interface RoleSpec {
  name: string
  description: string
  boundaries: string[]
}

export interface BrainSpec {
  model: string
  strategy: 'react'
  max_iterations: number
}

export interface MemorySpec {
  short_term: boolean
  long_term: boolean
  rag_collection: string | null
}

export interface ToolBinding {
  name: string
  enabled: boolean
  config: Record<string, unknown>
}

export interface PromptSpec {
  system: string
  user_template: string
}

export interface AcceptanceCase {
  given: string
  when: string
  then: string
}

export type WorkflowNodeType = 'start' | 'think' | 'tool' | 'decision' | 'end'

export interface WorkflowNode {
  id: string
  type: WorkflowNodeType
  label: string
  tool_name?: string | null
  condition?: string | null
}

export interface WorkflowEdge {
  from_node: string
  to_node: string
  condition?: string | null
  label: string
}

export interface WorkflowSpec {
  enabled: boolean
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

/** The eight-slot blueprint compiled from one sentence. */
export interface AgentSpec {
  id: string
  version: number
  status: SpecStatus
  one_liner: string // 0
  role: RoleSpec // 1
  brain: BrainSpec // 2
  memory: MemorySpec // 3
  tools: ToolBinding[] // 4
  policies: string[] // 5
  prompts: PromptSpec // 6
  acceptance: AcceptanceCase[] // 7
  workflow: WorkflowSpec // 8 (optional)
  compiler_notes: string
  created_at?: string
  updated_at?: string
  created_by?: number | null
}

export interface SpecSummary {
  id: string
  name: string
  version: number
  status: SpecStatus
  one_liner: string
}

export interface SpecResponse {
  id: string
  name: string
  version: number
  status: SpecStatus
  spec: AgentSpec
}

export interface ToolCapability {
  name: string
  description: string
  input_schema: Record<string, unknown>
  verbs: string[]
  risk: ToolRisk
}

export interface RunStep {
  thought?: string | null
  action?: string | null
  observation?: string | null
}

export interface RunResponse {
  spec_id: string
  final: string
  trace: RunStep[]
}
